File: frontend/src/services/appointmentService.js
/** Author: Tuấn - HE204215
 * Author: DucTKH - HE204463
 * Created: 2026-06-01
 * Last Update: 2026-07-17
 * 
 * File này chứa các hàm gọi API (Service) liên quan đến quản lý Lịch hẹn và Hàng đợi
 * Hỗ trợ các thao tác như: lấy danh sách lịch hẹn, cập nhật trạng thái, xác nhận, check-in, và lấy dữ liệu dashboard
*/

import axiosClient from '../api/axiosClient'

export const appointmentService = {

  /* Hàm lấy toàn bộ lịch hẹn dùng cho trang hóa đơn / tìm kiếm lịch hẹn */
  getAllAppointments: (keyword = '') =>
    axiosClient.get('/v1/appointments/search', { params: keyword ? { keyword } : {} }),

  /* Hàm lấy danh sách tất cả các lịch hẹn trong ngày hôm nay */
  getTodayAppointments: () =>
    axiosClient.get('/v1/appointments/today'),

  /* Hàm cập nhật trạng thái của một lịch hẹn (ví dụ: WAITING, IN_PROGRESS, COMPLETED...) */
  updateStatus: (id, status) =>
    axiosClient.patch(`/v1/appointments/${id}/status`, null, { params: { status } }),

  /* Hàm xác nhận lịch hẹn và phân công cho một bác sĩ cụ thể (nếu có).
     reason bắt buộc khi đổi sang bác sĩ khác bác sĩ bệnh nhân đã đặt. */
  confirmAppointment: (id, doctorId, reason) =>
    axiosClient.patch(`/v1/appointments/${id}/confirm`, { doctorId: doctorId || null, reason: reason || null }),

  /* Hàm đánh dấu bệnh nhân đã có mặt tại phòng khám (Check-in) */
  checkInAppointment: (id) =>
    axiosClient.patch(`/v1/appointments/${id}/check-in`),

  /* Hàm tạo mới một lịch hẹn trực tiếp (Walk-in) cho bệnh nhân đến khám không đặt trước */
  createWalkInAppointment: (data) =>
    axiosClient.post('/v1/appointments/walk-in', data),

  /* Hàm lấy thông tin thống kê số liệu lịch hẹn cho Dashboard (tổng số ca, ca chờ, ca hoàn thành...).
     Có thể truyền date (YYYY-MM-DD) để xem thống kê của một ngày bất kỳ. */
  getDashboard: (date) =>
    axiosClient.get('/v1/appointments/dashboard', { params: date ? { date } : {} }),

  /* Hàm lấy danh sách lịch hẹn của một ngày bất kỳ (YYYY-MM-DD) cho lễ tân */
  getDailySchedule: (date) =>
    axiosClient.get('/v1/appointments/daily-schedule', { params: date ? { date } : {} }),

  /* Hàm lấy danh sách hàng đợi bệnh nhân dành riêng cho tài khoản Bác sĩ đang đăng nhập */
  getDoctorQueue: (date) =>
    axiosClient.get('/v1/appointments/doctor-queue', { params: date ? { date } : {} }),

  /* Hàm lấy danh sách khung giờ còn trống của 1 bác sĩ trong 1 ngày (YYYY-MM-DD)
     để bệnh nhân chọn khi đặt lịch */
  getAvailableSlots: (doctorId, date) =>
    axiosClient.get('/v1/appointments/available-slots', { params: { doctorId, date } }),

  /* Hàm đặt trước một lịch hẹn khám bệnh mới (từ phía bệnh nhân) */
  bookAppointment: (data) =>
    axiosClient.post('/v1/appointments/book', data),

  getMyAppointments: () =>
    axiosClient.get('/v1/appointments/my'),

  /* Huỷ lịch hẹn (bệnh nhân tự huỷ hoặc lễ tân huỷ), có thể kèm lý do */
  cancelAppointment: (id, reason) =>
    axiosClient.patch(`/v1/appointments/${id}/cancel`, reason ? { reason } : null),

  /* Bệnh nhân tự đổi giờ khám trong giới hạn cho phép */
  rescheduleAppointment: (id, newAppointmentTime) =>
    axiosClient.patch(`/v1/appointments/${id}/reschedule`, { newAppointmentTime }),

  // Le Thi Bich Ngan - HE204710 | Tạo: 18/07/2026
  // Chức năng: gọi API reassign (đổi bác sĩ/giờ tại quầy) — không gắn BR số cụ
  // thể, hỗ trợ UC-18. Dùng cho modal "Đổi lịch hẹn" trong AppointmentManagementPage.
  /* UC-18: Lễ tân/Manager đổi lịch hẹn (bác sĩ mới và/hoặc giờ mới) — không bắt
     buộc phải đổi cả hai, cần ít nhất 1 trong 3: doctorId / newAppointmentTime / reason.
     Khác rescheduleAppointment (bệnh nhân tự đổi, luôn về lại PENDING): reassign
     giữ nguyên trạng thái hiện tại vì lễ tân xử lý trực tiếp tại quầy. */
  reassignAppointment: (id, { doctorId, newAppointmentTime, reason } = {}) =>
    axiosClient.patch(`/v1/appointments/${id}/reassign`, {
      doctorId: doctorId || null,
      newAppointmentTime: newAppointmentTime || null,
      reason: reason || null,
    }),

  /* Lấy chi tiết 1 lịch hẹn theo id (dùng cho modal chi tiết / mở từ thông báo) */
  getById: (id) =>
    axiosClient.get(`/v1/appointments/${id}`),

  /* UC-13: gửi nhắc lịch thủ công cho 1 lịch hẹn (bỏ qua cửa sổ 24h) */
  sendReminder: (id) =>
    axiosClient.post(`/v1/appointments/${id}/send-reminder`),

  abandonExam: (appointmentId) =>
    axiosClient.post(`/v1/appointments/${appointmentId}/abandon`),
}

File: backend/src/main/java/com/ecms/service/AppointmentService.java
/**
 * Author: TuanTD
 * 
 * Interface Service định nghĩa các nghiệp vụ cốt lõi liên quan đến quản lý Lịch hẹn (Appointment)
 * Xử lý toàn bộ các luồng trạng thái lịch hẹn từ lúc đặt lịch (Online/Walk-in), Check-in, 
 * Phân công điều phối bác sĩ, cho đến thống kê dữ liệu báo cáo Dashboard.
 */

package com.ecms.service;

import com.ecms.dto.request.BookAppointmentRequest;
import com.ecms.dto.request.CancelAppointmentRequest;
import com.ecms.dto.request.ReassignAppointmentRequest;
import com.ecms.dto.request.RescheduleAppointmentRequest;
import com.ecms.dto.request.UpdateAppointmentNotesRequest;
import com.ecms.dto.request.WalkInAppointmentRequest;
import com.ecms.dto.response.AppointmentDashboardResponse;
import com.ecms.dto.response.AppointmentResponse;
import com.ecms.entity.AppointmentStatus;

import java.time.LocalDate;
import java.util.List;

public interface AppointmentService {

        /* Lấy danh sách tất cả các lịch hẹn trong ngày hôm nay */
        List<AppointmentResponse> getTodayAppointments();

        /* Cập nhật trạng thái thủ công cho một lịch hẹn cụ thể */
        AppointmentResponse updateAppointmentStatus(Long id, AppointmentStatus status);

        /* Lấy toàn bộ danh sách lịch hẹn lưu trữ trên hệ thống */
        List<AppointmentResponse> getAllAppointments();

        /* Lấy danh sách lịch sử lịch hẹn của một bệnh nhân cụ thể */
        List<AppointmentResponse> getMyAppointments(Long patientId);

        /* Xác nhận lịch hẹn trực tuyến và phân công bác sĩ phụ trách */
        AppointmentResponse confirmAppointment(Long id, Long doctorId);

        /**
         * UC-15: Check-in bệnh nhân tại quầy. Cấp số thứ tự hàng đợi theo bác sĩ + ngày
         * (BR-13) và lưu lại id nhân viên thực hiện check-in.
         */
        AppointmentResponse checkInAppointment(Long id, Long checkInByUserId);

        /* Tìm kiếm lịch hẹn linh hoạt theo từ khóa tìm kiếm (Keyword) */
        List<AppointmentResponse> searchAppointments(String keyword);

        /*
         * Lấy danh sách hàng đợi khám tổng quát của toàn bộ phòng khám theo ngày cụ thể
         */
        List<AppointmentResponse> getDoctorQueue(LocalDate date);

        /* Tạo mới một lịch hẹn trực tiếp tại quầy (Walk-in Appointment) */
        AppointmentResponse createWalkInAppointment(WalkInAppointmentRequest request);

        /* Đặt lịch khám trực tuyến (Online Booking) dành cho khách hàng/bệnh nhân */
        AppointmentResponse bookOnlineAppointment(BookAppointmentRequest request, String patientEmail);

        /* Lấy dữ liệu thống kê tổng hợp (Dashboard) của toàn phòng khám theo ngày */
        AppointmentDashboardResponse getDashboard(LocalDate date);

        /* Lấy danh sách hàng đợi khám riêng biệt của một Bác sĩ cụ thể trong ngày */
        List<AppointmentResponse> getDoctorQueue(LocalDate date, Long doctorId);

        /* Lấy dữ liệu thống kê (Dashboard) cá nhân của một Bác sĩ cụ thể theo ngày */
        AppointmentDashboardResponse getDashboard(LocalDate date, Long doctorId);

        /* Điều chuyển hoặc đổi lịch hẹn sang một bác sĩ hoặc khung giờ khác */
        AppointmentResponse reassignAppointment(Long id, ReassignAppointmentRequest request);

        /* Lấy danh sách lịch trình phân bổ làm việc chi tiết theo từng ngày */
        List<AppointmentResponse> getDailySchedule(LocalDate date);

        /*
         * Lấy danh sách lịch hẹn nằm trong một khoảng ngày cụ thể từ [startDate] đến
         * [endDate]
         */
        List<AppointmentResponse> getScheduleRange(LocalDate startDate, LocalDate endDate);

        /**
         * Huỷ lịch hẹn. isPatientSelf=true áp dụng BR-05 (không huỷ trễ hơn 1h trước
         * giờ khám) và kiểm tra quyền sở hữu lịch hẹn.
         */
        AppointmentResponse cancelAppointment(Long id, CancelAppointmentRequest request, String actingUserEmail,
                        boolean isPatientSelf);

        /** Bệnh nhân tự đổi giờ khám (giữ nguyên bác sĩ) trong giới hạn cho phép */
        AppointmentResponse reschedulePatientAppointment(Long id, RescheduleAppointmentRequest request,
                        String patientEmail);

        /** Lễ tân ghi chú thêm cho lịch hẹn, không đổi trạng thái */
        AppointmentResponse updateAppointmentNotes(Long id, UpdateAppointmentNotesRequest request);

        /** Lấy chi tiết 1 lịch hẹn theo id (dùng cho modal chi tiết). */
        AppointmentResponse getAppointmentById(Long id);

        /**
         * UC-13: Gửi nhắc lịch cho 1 lịch hẹn — gửi email cho bệnh nhân, đặt
         * reminderSent=true và tạo thông báo cho Lễ tân. Dùng cho cả cron job
         * lẫn endpoint nhắc thủ công (bỏ qua điều kiện cửa sổ 24h).
         */
        AppointmentResponse sendReminder(Long id);
}